packages/docusaurus-theme-common/src/utils/docsUtils.tsx

```diff
@@ -0,0 +1,186 @@
+export const isDocsPluginEnabled: boolean = !!useAllDocsData;
+
+// Using a Symbol because null is a valid  context value (a doc can have no sidebar)
+// Inspired by https://github.com/jamiebuilds/unstated-next/blob/master/src/unstated-next.tsx
+const EmptyContextValue: unique symbol = Symbol('EmptyContext');
+
+const DocsVersionContext = createContext<
+  PropVersionMetadata | typeof EmptyContextValue
+>(EmptyContextValue);
+
+export function DocsVersionProvider({
+  children,
+  version,
+}: {
+  children: ReactNode;
+  version: PropVersionMetadata | typeof EmptyContextValue;
+}): JSX.Element {
+  return (
+    <DocsVersionContext.Provider value={version}>
+      {children}
+    </DocsVersionContext.Provider>
+  );
+}
+
+export function useDocsVersion(): PropVersionMetadata {
+  const version = useContext(DocsVersionContext);
+  if (version === EmptyContextValue) {
+    throw new Error('This hook requires usage of <DocsVersionProvider>');
+  }
+  return version;
+}
+
+export function useDocById(id: string): PropVersionDoc;
+export function useDocById(id: string | undefined): PropVersionDoc | undefined;
+export function useDocById(id: string | undefined): PropVersionDoc | undefined {
+  const version = useDocsVersion();
+  if (!id) {
+    return undefined;
+  }
+  const doc = version.docs[id];
+  if (!doc) {
+    throw new Error(`no version doc found by id=${id}`);
+  }
+  return doc;
+}
+
+const DocsSidebarContext = createContext<
+  PropSidebar | null | typeof EmptyContextValue
+>(EmptyContextValue);
+
+export function DocsSidebarProvider({
+  children,
+  sidebar,
+}: {
+  children: ReactNode;
+  sidebar: PropSidebar | null;
+}): JSX.Element {
+  return (
+    <DocsSidebarContext.Provider value={sidebar}>
+      {children}
+    </DocsSidebarContext.Provider>
+  );
+}
+
+export function useDocsSidebar(): PropSidebar | null {
+  const sidebar = useContext(DocsSidebarContext);
+  if (sidebar === EmptyContextValue) {
+    throw new Error('This hook requires usage of <DocsSidebarProvider>');
+  }
+  return sidebar;
+}
+
+// Use the components props and the sidebar in context
+// to get back the related sidebar category that we want to render
+export function findSidebarCategory(
+  sidebar: PropSidebar,
+  predicate: (category: PropSidebarItemCategory) => boolean,
+): PropSidebarItemCategory | undefined {
+  // eslint-disable-next-line no-restricted-syntax
+  for (const item of sidebar) {
+    if (item.type === 'category') {
+      if (predicate(item)) {
+        return item;
+      } else {
+        const subItem = findSidebarCategory(item.items, predicate);
+        if (subItem) {
+          return subItem;
+        }
+      }
+    }
+  }
+  return undefined;
+}
+
+// If a category card has no link => link to the first subItem having a link
+export function findFirstCategoryLink(
+  item: PropSidebarItemCategory,
+): string | undefined {
+  if (item.href) {
+    return item.href;
+  }
+
+  // eslint-disable-next-line no-restricted-syntax
```

Comment:
This is something I need to check

![image](https://user-images.githubusercontent.com/749374/143619405-e12aa9a2-5277-47b5-a9bd-4e1322f2f44a.png)

Reply:
Airbnb guide is extremely frustrating in this regard. They assumed all for-of loops are polyfilled. There have been numerous arguments that it should be accepted now. However it seems that this loop never reaches the second iteration...